### ISSUE TITLE
Support for in-document SVG content

### ISSUE THREAD

**Issue by Oblomov** [asciidoctor/asciidoctor#3748]
Hello all,

my understanding is that it's not currently possible to have an inline SVG in an `.adoc`; at best, under appropriate safety conditions, one may inline an _external_ SVG. What I'm interested in instead would be a way to do something like the following (pure speculation on the syntax, of course):

```
[[someid]]
.caption content
[image]
+++
<svg>blah blah </svg>
+++
```

which would be equivalent to

```
[[someid]]
.caption content
[%inline]
image::someexternal.svg[]
```
with the `<svg>blah blah </svg>`  in `someexternal.svg`.

I've found a partial workaround with custom blocks that do raw passthrough, but the obvious issue is that I then need to recreate also all of the wrapping stuff (caption, div etc), which basically requires me to recreate `convert_image`.

Suggestions?

**Comment by mojavelinux** [comment]
You are correct this is not built-in functionality. This would make a great block extension. You can define a block named `svg`, then in the extension generate an image block that passes the SVG data through via the target (perhaps base 64 encoded)...or the extension could generate a passthrough block, but that's not as ideal for reasons you mentioned.

**Comment by mojavelinux** [comment]
I almost wonder if this would fit with asciidoctor-diagram. You could think of the raw SVG as the source of a "diagram", and it would handle all the necessary details of translating it to an image block.

**Comment by Oblomov** [comment]
Hello @mojavelinux and thanks for your replies.

I did try creating an image block with a data-uri: of the image content, but I wasn't successful. With the default settings, it would complain about `allow-uri-read` being disabled, and after enabling it still failed with the generic warning: `asciidoctor: WARNING: could not retrieve contents of SVG at URI: <mybase64data>` (without an explanation or hint about the reason why). The relevant part of the block processing extension is
```
    attrs['target'] = 'data:image/svg+xml;base64,' + Base64::strict_encode64(svg)
    attrs['format'] = 'svg'
    attrs['inline-option'] = true
    create_image_block parent, attrs
```
but maybe I'm  doing something wrong (still, the need to enable `allow-uri-read` is a bit annoying, although I'm sure there's security reasons for this).

As an aside, to make it work provisionally I'm doing something horrible: live-patching the `convert_image` method to change the first `if` into `if node.content? ; img = node.content ; elsif`, and then using
```
    block = create_block parent, :image, [svg], attrs, content_model: :raw
    if title
      block.title = title
      block.assign_caption (attrs.delete 'caption'), 'figure'
    end
    block
```
to generate the image block. Obviously this is just a horrible temporary hack until the effect can be achieved through standard means, so don't get offended by the yuckiness of it ;-)

Concerning `asciidoctor-diagram` (ADD), my final use-case is actually something similar: the SVG isn't hand-coded, but gets generated programmatically with an ASCII to SVG conversion that uses some Ruby code I ported over from [Markdeep](https://casual-effects.com/markdeep/)'s JavScript implementation. My intention was to propose the inclusion of the final thing into ADD after it was properly cleaned up, and thus solve the two main issues I have with the way ADD works now: dependency on external programs (solved by the Ruby conversion code), and generation+caching of files instead of a direct embedding: the latter is the part where I got stuck, when I discovered the core didn't support it out of the box.

**Comment by Mogztter** [comment]
I guess this proposal https://github.com/asciidoctor/asciidoctor/issues/3698 would fix this issue since it would allow us to create an image block with the actual SVG contents (and the converter will use it directly and not try to read/retrieve the contents again).

**Comment by mojavelinux** [comment]
The problem you're facing is with the code path when the inline option is set and the target is a data URI. The logic is assuming this is a remote URL.

You can work around this problem by setting the interactive option instead. That will push the data URI into the data attribute of the `<object>` element. Here's what I was able to use to make it work:

```ruby
process do |parent, reader, attrs|
  svg_data = reader.lines.join ?\n
  target = %(data:image/svg+xml;base64,#{Base64::strict_encode64 svg_data})
  create_image_block parent, (attrs.merge 'target' => target, 'format' => 'svg', 'interactive-option' => '')
end
```

I can fix the code path in core so it base64 decodes the data URI. That will allow you to use the inline option and see the raw SVG source in the output document.

Guillaume is right that the fix for #3698 is a more universal solution to the problem. But I think this case is presenting a bug with the existing inline logic, so it's more narrow.

**Comment by Oblomov** [comment]
For the matter, the issue with `data:` is independent of the use of base64, even an unencoded `data:image/svg+xml,svg_data` has the same issue (except that the consequences on the output are even more catastrophic, of course). But the fact that `data:` URIs are treated like remote resources is definitely an issue that needs to be fixed, regardless of this specific application. (I mean, in some sense they are the “safest” of URIs, even safer than reading data from a separate local resource.)
Even this simple test case exhibits the issue:

```
.some test image
[%inline,format=svg]
image::data:image/svg+xml,<svg></svg>[]
```
(And it's actually two issues, because `format=svg` shouldn't even be necessary in this case).

I think fixing the handling of `data:` URI would essentially solve my specific problem, since then a custom block to do inline images directly (generated or user-produced) would be trivial.

I wonder if it would actually fix #3698 too (and sorry for not noticing that issue earlier, I was a bit transfixed on SVG), since then it would suffice to set the image node target to a `data:` URI with the contents in bytes (possibly base64-encoded).

**Comment by Oblomov** [comment]
BTW @Mogztter yes, that's exactly what I did with my patching: if the node has a content, then the content gets used, otherwise the standard processing applies. I'm convinced that proper handling of `data:` URIs would fix not only my issue, but also allow what you ask for in #3698, although possibly in a rather clumsy way (unless `create_image_block` is extended to allow an additional optional parameter holding the content, if target is missing (and doing the conversion to data URI behind the scene). Or something like that.

**Comment by mojavelinux** [comment]
> But the fact that data: URIs are treated like remote resources is definitely an issue that needs to be fixed, regardless of this specific application.

data: URIs are not treated like remote resources normally. It's just this code path that gets missed. And it's because it is an unanticipated / untested manipulation of the target by an extension.

**Comment by mojavelinux** [comment]
> even an unencoded data:image/svg+xml,svg_data has the same issue

I don't recommend defining the data URI for an SVG this way as it gets us into escaping rules for the tags and newlines, and that's just messy.

**Comment by mojavelinux** [comment]
I'll file a follow-up issue to resolve this problem.

**Comment by mojavelinux** [comment]
> even an unencoded data:image/svg+xml,svg_data has the same issue

It looks like I'll be able to support this scenario after all...it's easy to check for the base64, prefix and only decode if that is present.

**Comment by mojavelinux** [comment]
> it is an unanticipated / untested manipulation of the target by an extension.

As you pointed out, it's possible to create this scenario using bare AsciiDoc. It's the inline SVG code path that's the problem.

**Comment by Oblomov** [comment]
> > even an unencoded data:image/svg+xml,svg_data has the same issue
> 
> I don't recommend defining the data URI for an SVG this way as it gets us into escaping rules for the tags and newlines, and that's just messy.

While that's true, we cannot guarantee that the user won't do it. Probably for unencoded data URIs AsciiDoctor should do the encoding itself before placing the content inside an attribute.

**Comment by mojavelinux** [comment]
> While that's true, we cannot guarantee that the user won't do it. 

As I said, that's just a recommendation. I can still support it.

> Probably for unencoded data URIs AsciiDoctor should do the encoding itself before placing the content inside an attribute.

It's the input that would be the issue, not the output. But since the target of an image macro accepts any characters except newlines, it's not as messy as I first thought.

**Comment by michaeljfazio** [comment]
Linked back to [this discussion](https://github.com/asciidoctor/asciidoctor-diagram/issues/213) due to some common goals in making use of SVG for more flexible diagram support using drawio.
